test/framework/clusterctl/clusterctl_helpers.go

```diff
@@ -127,28 +127,50 @@ func InitManagementClusterAndWatchControllerLogs(ctx context.Context, input Init
 
 // UpgradeManagementClusterAndWaitInput is the input type for UpgradeManagementClusterAndWait.
 type UpgradeManagementClusterAndWaitInput struct {
-	ClusterProxy         framework.ClusterProxy
-	ClusterctlConfigPath string
-	ClusterctlVariables  map[string]string
-	Contract             string
-	LogFolder            string
+	ClusterProxy              framework.ClusterProxy
+	ClusterctlConfigPath      string
+	ClusterctlVariables       map[string]string
+	Contract                  string
+	CoreProvider              string
+	BootstrapProviders        []string
+	ControlPlaneProviders     []string
+	InfrastructureProviders   []string
+	IPAMProviders             []string
+	RuntimeExtensionProviders []string
+	LogFolder                 string
 }
 
 // UpgradeManagementClusterAndWait upgrades provider a management cluster using clusterctl, and waits for the cluster to be ready.
 func UpgradeManagementClusterAndWait(ctx context.Context, input UpgradeManagementClusterAndWaitInput, intervals ...interface{}) {
 	Expect(ctx).NotTo(BeNil(), "ctx is required for UpgradeManagementClusterAndWait")
 	Expect(input.ClusterProxy).ToNot(BeNil(), "Invalid argument. input.ClusterProxy can't be nil when calling UpgradeManagementClusterAndWait")
 	Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterctlConfigPath must be an existing file when calling UpgradeManagementClusterAndWait")
-	Expect(input.Contract).ToNot(BeEmpty(), "Invalid argument. input.Contract can't be empty when calling UpgradeManagementClusterAndWait")
+	// Check if the user want a custom upgrade
+	isCustomUpgrade := input.CoreProvider != "" ||
+		len(input.BootstrapProviders) > 0 ||
+		len(input.ControlPlaneProviders) > 0 ||
+		len(input.InfrastructureProviders) > 0 ||
+		len(input.IPAMProviders) > 0 ||
+		len(input.RuntimeExtensionProviders) > 0
+
+	Expect((input.Contract != "" && !isCustomUpgrade) || (input.Contract == "" && isCustomUpgrade)).To(BeTrue(), `Invalid argument. Either the input.Contract parameter or at least one of the following providers has to be set:
```

Comment:
```suggestion
	Expect((input.Contract != "" && isCustomUpgrade) || (input.Contract == "" && !isCustomUpgrade)).To(BeTrue(), `Invalid argument. Either the input.Contract parameter or at least one of the following providers has to be set:
```

I think that's the correct validation.

Essentially:
* either contract
* or one of the provider fields

can be set. Both unset is an error and both set is an error

Reply:
At least that's my interpretation of:
```go
	if ua.contract == "" && !hasProviderNames {
		return errors.New("Either the --contract flag or at least one of the following flags has to be set: --core, --bootstrap, --control-plane, --infrastructure, --ipam, --extension")
	}
	if ua.contract != "" && hasProviderNames {
		return errors.New("The --contract flag can't be used in combination with --core, --bootstrap, --control-plane, --infrastructure, --ipam, --extension")
	}
```

Reply:
(same in client.go)

Reply:
I think the logic with expect is bit different we are checking when to pass and the other example check when to return error.

to use the same interpretation we may change also the`to(BeTrue())`

Reply:
Sorry my bad :/ You're right